client/executor/wasmtime/src/instance_wrapper.rs

```diff
@@ -136,62 +137,41 @@ fn extern_func(extern_: &Extern) -> Option<&Func> {
 	}
 }
 
+pub(crate) fn create_store(engine: &wasmtime::Engine, max_memory_size: Option<usize>) -> Store {
+	let limits = if let Some(max_memory_size) = max_memory_size {
+		wasmtime::StoreLimitsBuilder::new().memory_size(max_memory_size).build()
+	} else {
+		Default::default()
+	};
+
+	let mut store =
+		Store::new(engine, StoreData { limits, host_state: None, memory: None, table: None });
+	if max_memory_size.is_some() {
+		store.limiter(|s| &mut s.limits);
+	}
+	store
+}
+
 impl InstanceWrapper {
-	/// Create a new instance wrapper from the given wasm module.
-	pub fn new<H>(
-		module: &Module,
-		heap_pages: u64,
-		allow_missing_func_imports: bool,
+	pub(crate) fn new(
+		engine: &Engine,
+		instance_pre: &InstancePre<StoreData>,
 		max_memory_size: Option<usize>,
-	) -> Result<Self>
-	where
-		H: HostFunctions,
-	{
-		let limits = if let Some(max_memory_size) = max_memory_size {
-			wasmtime::StoreLimitsBuilder::new().memory_size(max_memory_size).build()
-		} else {
-			Default::default()
-		};
-
-		let mut store = Store::new(
-			module.engine(),
-			StoreData { limits, host_state: None, memory: None, table: None },
-		);
-		if max_memory_size.is_some() {
-			store.limiter(|s| &mut s.limits);
-		}
-
-		// Scan all imports, find the matching host functions, and create stubs that adapt arguments
-		// and results.
-		let imports = crate::imports::resolve_imports::<H>(
-			&mut store,
-			module,
-			heap_pages,
-			allow_missing_func_imports,
-		)?;
-
-		let instance = Instance::new(&mut store, module, &imports.externs)
-			.map_err(|e| Error::from(format!("cannot instantiate: {}", e)))?;
-
-		let memory = match imports.memory_import_index {
-			Some(memory_idx) => extern_memory(&imports.externs[memory_idx])
-				.expect("only memory can be at the `memory_idx`; qed")
-				.clone(),
-			None => {
-				let memory = get_linear_memory(&instance, &mut store)?;
-				if !memory.grow(&mut store, heap_pages).is_ok() {
-					return Err("failed top increase the linear memory size".into())
-				}
```

Comment:
Ah, good catch!

In general this should be harmless since the memory will grow anyway as necessary. (And with the current instantiation scheme that will only happen once, so it doesn't matter for performance since the memory is reused anyway.) That said, I *did* preserve this behavior for imported memories, so I should probably also do it for exported memories for consistency.

A side note: this code is somewhat... strange? Essentially the `heap_pages` which is passed to the executor *doesn't* actually define how many heap pages there should be, just how many *extra* heap pages should be allocated *in addition* to what's declared in the runtime blob. And AFAIK we don't actually control how many pages the runtime blob requests - it's whatever `rustc` decides to put in as the default. So overall what we're doing is `allocate(some_random_minimum_number_of_heap_pages_rustc_put_in_our_wasm_blob + extra_heap_pages)`, which doesn't make much sense to me (maybe @pepyakin remembers why it was done that way?). It'd probably make more sense to use the `heap_pages` as a minimum instead, or just get rid of it completely and leave the initial size of the memory to be equal to what's actually requested by the WASM blob. (But I don't really want to change the behavior in this PR, so I'll just add extra tests and restore the original behavior.)